File: frontend/src/services/api/evidence.ts
/** Investigation evidence and its provenance (V9). */

import { api } from "./client";

/** What kind of claim a piece of evidence is. */
export type EvidenceOrigin =
  | "observed"
  | "derived"
  | "reported"
  | "analytic"
  | "simulated";

/** What the backend can actually promise about the stored record. */
export type EvidenceIntegrity = "append_only" | "write_once" | "mutable";

export interface ApiProvenance {
  provider: string;
  /** Typed pointer back to the source row: `"<type>:<id>"`. */
  sourceRef: string;
  origin: EvidenceOrigin;
  integrity: EvidenceIntegrity;
  /**
   * False for a mutable source: the content behind this item can change after
   * a decision was taken on it, and only the digest would show that it had.
   */
  tamperEvidentAtRest: boolean;
  /** When the thing happened. Null when the source genuinely does not know. */
  observedAt: string | null;
  /** When AEGISX recorded it. Never inferred from `observedAt`. */
  collectedAt: string;
  confidence: number | null;
  /** What the confidence measures. Always present when there is one. */
  confidenceBasis: string | null;
  incidentRef: string | null;
  eventRef: string | null;
  isSynthetic: boolean;
  extra: Record<string, unknown>;
}

export interface ApiEvidenceItem {
  evidenceId: string;
  kind: string;
  title: string;
  content: Record<string, unknown>;
  contentDigest: string;
  /** The text looks like an attempt to steer a language model. */
  containsInjectionAttempt: boolean;
  provenance: ApiProvenance;
}

export interface ApiDegradedProvider {
  provider: string;
  status: string;
  reason: string | null;
}

export interface ApiEvidenceSet {
  incidentId: string;
  manifestDigest: string;
  total: number;
  countsByKind: Record<string, number>;
  countsByOrigin: Record<string, number>;
  injectionFlagged: string[];
  /**
   * Providers that could not answer. Rendered separately from an empty item
   * list on purpose: "no evidence" and "we could not ask" are different facts.
   */
  degradedProviders: ApiDegradedProvider[];
  filters: Record<string, unknown>;
  items: ApiEvidenceItem[];
}

export async function fetchIncidentEvidence(
  incidentId: string,
  params?: { kind?: string; provider?: string },
): Promise<ApiEvidenceSet> {
  const { data } = await api.get<ApiEvidenceSet>(
    `/incidents/${incidentId}/evidence`,
    { params },
  );
  return data;
}

export async function fetchEvidenceItem(
  incidentId: string,
  evidenceId: string,
): Promise<ApiEvidenceItem> {
  const { data } = await api.get<ApiEvidenceItem>(
    `/incidents/${incidentId}/evidence/${evidenceId}`,
  );
  return data;
}

File: backend/app/schemas/evidence.py
"""Evidence API schemas.

Response shapes only. There is deliberately no request body anywhere in this
module: the evidence API is read-only, so there is nothing a caller can send
that would change what an item says about itself.
"""

from __future__ import annotations

from typing import Any

from pydantic import Field

from app.schemas.common import CamelModel


class ProvenanceRead(CamelModel):
    """Where one piece of evidence came from, and what it is worth."""

    provider: str
    #: Typed pointer back to the row: ``"<type>:<id>"``.
    source_ref: str
    #: observed / derived / reported / analytic / simulated.
    origin: str
    #: append_only / write_once / mutable - what the platform can actually
    #: promise about the stored row, not what it would like to claim.
    integrity: str
    #: False for a mutable source: the content behind this item can change
    #: after a decision was made on it.
    tamper_evident_at_rest: bool
    observed_at: str | None = None
    collected_at: str
    confidence: float | None = None
    #: What the confidence number measures. Always present when there is one.
    confidence_basis: str | None = None
    incident_ref: str | None = None
    event_ref: str | None = None
    is_synthetic: bool = False
    extra: dict[str, Any] = Field(default_factory=dict)


class EvidenceItemRead(CamelModel):
    evidence_id: str
    kind: str
    title: str
    #: The normalised facts. Not scrubbed: an injected command line *is* the
    #: evidence, and hiding it from an analyst would hide the attack.
    content: dict[str, Any] = Field(default_factory=dict)
    content_digest: str
    #: True when the text looks like an attempt to steer a language model.
    contains_injection_attempt: bool = False
    provenance: ProvenanceRead


class DegradedProvider(CamelModel):
    provider: str
    status: str
    reason: str | None = None


class EvidenceSetRead(CamelModel):
    incident_id: str
    #: One digest over every item's identity and content. Stable while the
    #: evidence is; changes if anything is added, removed or altered.
    manifest_digest: str
    total: int
    counts_by_kind: dict[str, int] = Field(default_factory=dict)
    counts_by_origin: dict[str, int] = Field(default_factory=dict)
    injection_flagged: list[str] = Field(default_factory=list)
    #: Providers that could not answer. Separate from an empty item list, so
    #: "no evidence" and "we could not ask" never render the same way.
    degraded_providers: list[DegradedProvider] = Field(default_factory=list)
    filters: dict[str, Any] = Field(default_factory=dict)
    items: list[EvidenceItemRead] = Field(default_factory=list)
